### PR TITLE
[autoparallel] implemented linear projection strategy generator

### PR DESCRIPTION
**Pull request by FrankLeeeee**
This PR implemented the strategy generator for linear operation. This PR can be summarized as follows:

1. create a module called strategy which stores the related classes and functions for sharding strategy and generators
2. implemented the generators for linear operations

### CODE REVIEW

File: colossalai/auto_parallel/solver/sharding_strategy.py [hunk 1]
```diff
@@ -122,9 +126,9 @@ class ShardingStrategy_V2:
     """
     name: str
     sharding_specs: Dict[OperationData, ShardingSpec] = None
-    compute_cost: TrainCycleItem = None
-    communication_cost: TrainCycleItem = None
-    memory_cost: TrainCycleItem = None
+    compute_cost: TrainCycleItem[int] = None
```
Comment (YuliangLiu0306) [inline colossalai/auto_parallel/solver/sharding_strategy.py:129]: I am not sure about whether TrainCyclyItem object is subscriptable or not

Comment (FrankLeeeee) [reply colossalai/auto_parallel/solver/sharding_strategy.py]: Fixed.

File: colossalai/auto_parallel/solver/strategy/matmul_strategy_generator.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+            },
+            self.operation_data_mapping['bias']: {
+                -1: [mesh_dim_1]
+            },
+            self.operation_data_mapping['output']: {
+                0: [mesh_dim_0],
+                -1: [mesh_dim_1]
+            },
+        }
+        communication_actions = {}
```
Comment (YuliangLiu0306) [inline colossalai/auto_parallel/solver/strategy/matmul_strategy_generator.py:117]: For training mission, all reduce communication is required during computing the gradient of parameters, and gradient of activation.

File: colossalai/auto_parallel/solver/strategy/matmul_strategy_generator.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+                0: [mesh_dim_0, mesh_dim_1]
+            },
+            self.operation_data_mapping['other']: {},
+            self.operation_data_mapping['bias']: {},
+            self.operation_data_mapping['output']: {
+                0: [mesh_dim_0, mesh_dim_1]
+            },
+        }
+        communication_actions = {
+            self.operation_data_mapping['other']: [
```
Comment (YuliangLiu0306) [inline colossalai/auto_parallel/solver/strategy/matmul_strategy_generator.py:235]: For better performance, flatten device mesh should be used here.

File: colossalai/auto_parallel/solver/strategy/matmul_strategy_generator.py (hunk 2 above)
Comment (FrankLeeeee) [reply colossalai/auto_parallel/solver/strategy/matmul_strategy_generator.py]: done.

File: colossalai/auto_parallel/solver/strategy/matmul_strategy_generator.py (hunk 3 above)
Comment (FrankLeeeee) [reply colossalai/auto_parallel/solver/strategy/matmul_strategy_generator.py]: updated with comm spec.

**Review by YuliangLiu0306 — APPROVED** [review]

**Pull request merged**